src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -0,0 +1,432 @@
+        for (Map.Entry<Integer, Milestone> entry : milestonesInJob.entrySet()) {
+            if (entry.getKey().equals(ordinal)) {
+                continue;
+            }
+            Milestone milestone2 = entry.getValue();
+            // The build is passing a milestone, so it's not visible to any previous milestone
+            if (milestone2.wentAway(r)) {
+                // Ordering check
+                if(milestone2.ordinal != ordinal - 1) {
+                    throw new AbortException(String.format("Unordered milestone. Found ordinal %s but %s was expected.", ordinal, milestone2.ordinal + 1));
+                }
+                // Cancel older builds (holding or waiting to enter)
+                cancelOldersInSight(milestone2, r);
+            }
+        }
+
+        // checking order
+        if (milestone.lastBuild != null && r.getNumber() < milestone.lastBuild) {
+            // cancel if it's older than the last one passing this milestone
+            cancel(context, milestone.lastBuild);
+        } else {
+            // It's in-order, proceed
+            milestone.pass(context, r);
+        }
+        cleanUp(job, jobName);
+        save();
+    }
+
+    private static synchronized void exit(Run<?,?> r) {
+        load();
+        LOGGER.log(Level.FINE, "exit {0}: {1}", new Object[] {r, getMilestonesByOrdinalByJob()});
+        Job<?,?> job = r.getParent();
+        String jobName = job.getFullName();
+        Map<Integer, Milestone> milestonesInJob = getMilestonesByOrdinalByJob().get(jobName);
+        if (milestonesInJob == null) {
+            return;
+        }
+        boolean modified = false;
+        for (Milestone milestone : milestonesInJob.values()) {
+            if (milestone.wentAway(r)) {
+                modified = true;
+                cancelOldersInSight(milestone, r);
+            }
+        }
+        if (modified) {
+            cleanUp(job, jobName);
+        }
+
+        // Clean non-existing milestones
+        if (r instanceof FlowExecutionOwner.Executable) {
+            Integer lastMilestoneOrdinal = getLastOrdinalInBuild(r);
+            if (lastMilestoneOrdinal == null) {
+                return;
+            }
+            Milestone m = getFirstWithoutInSight(milestonesInJob);
+            while (m != null && milestonesInJob.size() - 1 > lastMilestoneOrdinal) {
+                modified = true;
+                milestonesInJob.remove(m.ordinal);
+                m = getFirstWithoutInSight(milestonesInJob);
+            }
+            if (milestonesInJob.isEmpty()) {
+                modified = true;
+                getMilestonesByOrdinalByJob().remove(jobName);
+            }
+        }
+
+        if (modified) {
+            save();
+        }
+    }
+
+    @CheckForNull
+    private static Integer getLastOrdinalInBuild(Run<?, ?> r) {
+        int lastMilestoneOrdinal = 0;
+        FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) r).asFlowExecutionOwner();
```

Comment:
:ant: Better to take `FlowExecutionOwner.Executable` as the parameter type so the cast immediately follows the `instanceof` check.